File: src/go/uav.go
package PhantasyIslandGoRemoteControl

import (
	"time"
)

// UAV 此类是到 UAV 库的适配器，是对AirplaneManager的wrapper
type UAV struct {
	Airs *AirplaneManager
}

func NewUAV() *UAV {
	u := &UAV{
		Airs: GetAirplaneManager(),
	}
	u.Airs.Flush()
	u.Airs.Start()
	u.Airs.Flush()
	return u
}

func (u *UAV) Sleep(seconds float64) {
	time.Sleep(time.Duration(seconds * float64(time.Second)))
}

func (u *UAV) Destroy() {
	// 实现销毁逻辑
}

func (u *UAV) AddUav(port string) {
	u.Airs.GetAirplane(port)
}

func (u *UAV) P(port string) *AirplaneController {
	return u.Airs.GetAirplane(port)
}

func (u *UAV) Land(port string) {
	u.P(port).Land()
}

func (u *UAV) Emergency(port string) {
	u.P(port).Stop()
}

func (u *UAV) Takeoff(port string, high int) {
	u.P(port).Takeoff(high)
}

func (u *UAV) Up(port string, distance int) {
	u.P(port).Up(distance)
}

func (u *UAV) Down(port string, distance int) {
	u.P(port).Down(distance)
}

func (u *UAV) Forward(port string, distance int) {
	u.P(port).Forward(distance)
}

func (u *UAV) Back(port string, distance int) {
	u.P(port).Back(distance)
}

func (u *UAV) Left(port string, distance int) {
	u.P(port).Left(distance)
}

func (u *UAV) Right(port string, distance int) {
	u.P(port).Right(distance)
}

func (u *UAV) Goto(port string, x int, y int, h int) {
	u.P(port).Goto(x, y, h)
}

func (u *UAV) Flip(port string, direction string) {
	p := u.P(port)
	switch direction {
	case "f":
		p.FlipForward()
	case "b":
		p.FlipBack()
	case "r":
		p.FlipRight()
	case "l":
		p.FlipLeft()
	}
}

func (u *UAV) Rotate(port string, degree int) {
	u.P(port).Rotate(degree)
}

func (u *UAV) Cw(port string, degree int) {
	u.P(port).Cw(degree)
}

func (u *UAV) Ccw(port string, degree int) {
	u.P(port).Ccw(degree)
}

func (u *UAV) Speed(port string, speed int) {
	u.P(port).Speed(speed)
}

func (u *UAV) High(port string, high int) {
	u.P(port).High(high)
}

func (u *UAV) Led(port string, r, g, b int) {
	u.P(port).Led(r, g, b)
}

func (u *UAV) Bln(port string, r, g, b int) {
	u.P(port).Bln(r, g, b)
}

func (u *UAV) Rainbow(port string, r, g, b int) {
	u.P(port).Rainbow(r, g, b)
}

func (u *UAV) Mode(port string, mode int) {
	u.P(port).AirplaneMode(mode)
}

func (u *UAV) Stop(port string) {
	u.P(port).Stop()
}

func (u *UAV) Hover(port string) {
	u.P(port).Hover()
}

File: src/ts/uav.ts
import {AirplaneManager, getAirplaneManager} from './airplaneManager';
import {AirplaneController} from './controlCommand';

export class UAV {
    /**
     * 此类是到 UAV 库的适配器，是对 AirplaneManager 的 wrapper
     */
    airs: AirplaneManager = getAirplaneManager();

    constructor() {
        this.init();
    }

    private async init() {
        await this.airs.flush();
        await this.airs.start();
        await this.airs.flush();
    }

    async sleep(time: number) {
        /** sleep 单位 秒 */
        return new Promise(resolve => setTimeout(resolve, time * 1000));
    }

    destroy() {
        // this.airs.destroy()
    }

    async addUav(port: string) {
        /** 添加（注册）无人机 */
        return this.airs.getAirplane(port);
    }

    p(port: string): AirplaneController | undefined {
        return this.airs.getAirplane(port);
    }

    async land(port: string) {
        /** 降落 */
        const uav = this.p(port);
        if (uav) return uav.land();
    }

    async emergency(port: string) {
        const uav = this.p(port);
        if (uav) return uav.stop();
    }

    async takeoff(port: string, high: number) {
        /** 起飞到指定高度 单位cm */
        const uav = this.p(port);
        if (uav) return uav.takeoff(high);
    }

    async up(port: string, distance: number) {
        /** 上升指定距离 单位cm */
        const uav = this.p(port);
        if (uav) return uav.up(distance);
    }

    async down(port: string, distance: number) {
        /** 下降指定距离 单位cm */
        const uav = this.p(port);
        if (uav) return uav.down(distance);
    }

    async forward(port: string, distance: number) {
        /** 前进指定距离 单位cm */
        const uav = this.p(port);
        if (uav) return uav.forward(distance);
    }

    async back(port: string, distance: number) {
        /** 后退指定距离 单位cm */
        const uav = this.p(port);
        if (uav) return uav.back(distance);
    }

    async left(port: string, distance: number) {
        /** 左移指定距离 单位cm */
        const uav = this.p(port);
        if (uav) return uav.left(distance);
    }

    async right(port: string, distance: number) {
        /** 右移指定距离 单位cm */
        const uav = this.p(port);
        if (uav) return uav.right(distance);
    }

    async goto(port: string, x: number, y: number, h: number) {
        /** 移动到指定坐标处 */
        const uav = this.p(port);
        if (uav) return uav.goto(x, y, h);
    }

    async flip(port: string, direction: string) {
        /**
         * flip 函数用于控制无人机翻滚
         * @param port
         * @param direction 翻滚方向（f前 b后 l左 r右）
         */
        const uav = this.p(port);
        if (!uav) return;
        if (direction === 'f') {
            return uav.flipForward();
        } else if (direction === 'b') {
            return uav.flipBack();
        } else if (direction === 'r') {
            return uav.flipRight();
        } else if (direction === 'l') {
            return uav.flipLeft();
        }
    }

    async rotate(port: string, degree: number) {
        /** 顺时旋转指定角度 */
        const uav = this.p(port);
        if (uav) return uav.rotate(degree);
    }

    async cw(port: string, degree: number) {
        /** 顺时针旋转指定角度 */
        const uav = this.p(port);
        if (uav) return uav.cw(degree);
    }

    async ccw(port: string, degree: number) {
        /** 逆时针旋转指定角度 */
        const uav = this.p(port);
        if (uav) return uav.ccw(degree);
    }

    async speed(port: string, speed: number) {
        /** 设置飞行速度 */
        const uav = this.p(port);
        if (uav) return uav.speed(speed);
    }

    async high(port: string, high: number) {
        /** 移动到指定高度处 */
        const uav = this.p(port);
        if (uav) return uav.high(high);
    }

    async led(port: string, r: number, g: number, b: number) {
        /** 设置无人机led色彩 */
        const uav = this.p(port);
        if (uav) return uav.led(r, g, b);
    }

    async bln(port: string, r: number, g: number, b: number) {
        /** 设置无人机led呼吸灯色彩 */
        const uav = this.p(port);
        if (uav) return uav.bln(r, g, b);
    }

    async rainbow(port: string, r: number, g: number, b: number) {
        /** 设置无人机led彩虹色彩 */
        const uav = this.p(port);
        if (uav) return uav.rainbow(r, g, b);
    }

    async mode(port: string, m: number) {
        /** 设置无人机飞行模式 */
        const uav = this.p(port);
        if (uav) return uav.airplaneMode(m);
    }

    async stop(port: string) {
        /** 停桨 */
        const uav = this.p(port);
        if (uav) return uav.stop();
    }

    async hover(port: string) {
        /** 悬停 */
        const uav = this.p(port);
        if (uav) return uav.hover();
    }
}
